[DROP-IN] Add ability for interested UIComponents to observe changes to map styles.

tomaszrybakiewicz opened the pull request:
Closes [#1568](https://github.com/mapbox/navigation-sdks/issues/1568)

### Description

- Introduced MapStyleLoader class that's responsible for: 
  - loading initial map Style
  - reloading map Style on either `NavigationViewOptions.mapStyleUriNight` or
 `NavigationViewOptions.mapStyleUriDay` _flow_ changes
  - registers itself as `OnStyleLoadedListener` and updates `loadedMapStyle` StateFlow value
 every time new Style is fully loaded
- Updated UIBinders to observe loadedMapStyle StateFlow and reload RouteLineComponent, RouteArrowComponent, RoadNameLabelComponent and ManeuverComponent
- Updated MapLayoutCoordinator to call MapStyleLoader when loading default and custom map view

_MapStyleLoader class diagram_
![drop in mapstyleloader updated](https://user-images.githubusercontent.com/2678039/160913698-7f2a9524-950c-4a17-b265-54c2c3f79da4.png)


### Screenshots or Gifs

https://user-images.githubusercontent.com/2678039/160492299-ebbc40d5-b7ad-45c1-8177-bbf6e060f019.mp4

**Known issue**
Map doesn't recenter on current location after configuration change.

tomaszrybakiewicz commented on libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/component/roadlabel/RoadNameViewBinderTest.kt:
```diff
@@ -1,35 +0,0 @@
-package com.mapbox.navigation.dropin.component.roadlabel
```
Will re-add binder tests once I figure out how to test `reloadOnChange`  result.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt:
```diff
@@ -131,14 +132,6 @@ class DropInNavigationView @JvmOverloads constructor(
 
     override fun getLifecycle(): Lifecycle = viewLifecycleRegistry
 
-    private fun retrieveCurrentUiMode(): Int {
```
❤️

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/MapStyleLoader.kt:
```diff
@@ -0,0 +1,88 @@
+    private val mapboxMap: MapboxMap,
+    private val options: NavigationViewOptions,
+    private val loadedMapStyle: MutableStateFlow<Style?>,
+) : OnStyleLoadedListener, DefaultLifecycleObserver {
```
```suggestion
) : DefaultLifecycleObserver {
```

Small opinion here. Move the callback interface inside, doesn't need to be exposed on the top level

tomaszrybakiewicz replied:
Done

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/map/RoutePreviewMapBinder.kt:
```diff
@@ -25,12 +25,14 @@ internal class RoutePreviewMapBinder(
-                    lineOptions,
-                    navigationViewContext.viewModel.routesViewModel
-                )
+            reloadOnChange(navigationViewContext.loadedMapStyle) {
+                reloadOnChange(navigationViewContext.options.routeLineOptions) { lineOptions ->
```
Maybe not in this PR, but it would be nice to be able to combine these

```suggestion
            reloadOnChange(navigationViewContext.loadedMapStyle, navigationViewContext.options.routeLineOptions) { _, lineOptions ->

```

tomaszrybakiewicz replied:
Yeah... We could follow coroutines flow transform approach and introduce something similar to `collect(): Flow`.

tomaszrybakiewicz replied:
Done.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt:
```diff
@@ -81,12 +80,14 @@ class DropInNavigationView @JvmOverloads constructor(
-        }
-        binding.mapView.getMapboxMap().loadStyleUri(style)
+        lifecycle.addObserver(
+            MapStyleLoader(
+                context,
+                binding.mapView.getMapboxMap(),
+                navigationContext.options,
+                navigationContext.loadedMapStyle
+            )
```
Did you consider adding the `MapStyleLoader` instance to the `DropInNavigtationViewContext`

I think we should expose a class for downstream components to use. Instead of exposing the MutableStateFlow

tomaszrybakiewicz replied:
I did consider and initially tried approach. Unfortunately it lead to massive memory leak somewhere that took way too much time to hunt down. 
So I moved  'loadedMapStyle' StateFlow outside and refactored MapStyleLoader to ensure that it is only retained by the Lifecycle.

kmadsen replied:
Oh interesting.. Do we need to null out the `DropInNavigationViewContext` when the view is destroyed? Do you know where the leak is happening? I can help hunt it down

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/MapStyleLoader.kt:
```diff
@@ -0,0 +1,88 @@
+    private val context: Context,
+    private val mapboxMap: MapboxMap,
+    private val options: NavigationViewOptions,
+    private val loadedMapStyle: MutableStateFlow<Style?>,
```
```suggestion
```

I don't think we should pass a MutableStateFlow in this way. The drop-in-ui has domain specific classes that wrap the lower level state flows.

What do you think about using the MapStyleLoader as a dependency for the UIComponents https://github.com/mapbox/mapbox-navigation-android/pull/5621/files#r837761035

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewBinder.kt:
```diff
@@ -3,14 +3,20 @@ package com.mapbox.navigation.dropin.component.maneuver
 @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
-class ManeuverViewBinder : UIBinder {
+class ManeuverViewBinder(
+    private val loadedMapStyle: StateFlow<Style?>
```
I wonder how someone will override the `ManeuverViewBinder` 

```
with val maneuver: StateFlow<UIBinder?> get() = _maneuver.asStateFlow()
```

They will also need to have access to the stateflow given by `MapStyleLoader` 🤔 .. Wondering if this means we should expose map styles through `UICoordinator` => `UIComponent`

tomaszrybakiewicz replied:
`ManeuverCoordinator` does that. It observes `navigationViewContext.uiBinders.maneuver` and returns custom binder if needed.

kmadsen replied:
Resolving this one as addressed, we explored some options and they were not as good as this - or required large changes

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewBinder.kt:
```diff
@@ -20,6 +26,9 @@ class ManeuverViewBinder : UIBinder {
-        return ManeuverComponent(binding.maneuverView)
+        return reloadOnChange(loadedMapStyle) {
+            if (it != null) ManeuverComponent(binding.maneuverView, it)
+            else EmptyComponent()
```
small comment: We could have `reloadOnChange` take `null`. feels simpler to me considering the return type is `MapboxNavigationObserver`

```suggestion
            else null
```

``` kotlin
fun <T> reloadOnChange(value: Flow<T>, factory: (value: T) -> UIComponent): UIComponent =
    object : UIComponent() {
        private var childComponent: UIComponent? = null
```

kmadsen replied:
Otherwise instead of introducing the empty component, could it use NoOpMapboxNavigationObserver?

tomaszrybakiewicz replied:
Done

tomaszrybakiewicz commented:
@kmadsen Updated. Refactored MapStyleLoader, moved it to DropInNavigationViewContext and updated MapLayoutCoordinator to use it. (see UML diagram in PR description).
Let me know if there's anything else that needs to change.

kmadsen commented on libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/extensions/MapboxStyleExtensionsKtTest.kt:
```diff
@@ -0,0 +1,33 @@
+
+    // Map of (URI to EXPECTED_STYLE_ID)
+    private val fixtures: Map<String, String?> = mapOf(
+        NavigationStyles.NAVIGATION_NIGHT_STYLE to NavigationStyles.NAVIGATION_NIGHT_STYLE_ID,
```
Love this

kmadsen commented on qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/MapboxNavigationViewCustomizedActivity.kt:
```diff
@@ -86,14 +92,16 @@ class MapboxNavigationViewCustomizedActivity : AppCompatActivity() {
         // This demonstrates that you can customize views at any time. You can also reset to
         // the default views.
-        showCustomViews.observe(this) { showCustomViews ->
+        vm.showCustomViews.observe(this) { showCustomViews ->
```
will you please use a full name? IMO `viewModel` is acceptable but many argue that you should just spell it out

```suggestion
        customizedViewModel.showCustomViews.observe(this) { showCustomViews ->
```

`vm` is not a common abbreviation. just about every high level language style guide suggests camel case with full names. 

https://kotlinlang.org/docs/coding-conventions.html#choose-good-names
https://google.github.io/styleguide/javaguide.html#s5.2.7-local-variable-names
https://source.android.com/devices/architecture/hidl/code-style#naming

kmadsen replied:
Sorry if this was offensive, i do not mean it to be. I did some searching on style guides when I was wondering if i'm out of date with recent style guides. after searching, linked the findings

kmadsen reviewed (approved):
Great! Thanks for rebasing 🥇
